### PR TITLE
Restore / Backup prototype & Recipe Canonicalization & Raw JSON

### LINKED ISSUE #150: Relax minimum length for password to 6 (not 16)
User report

> One thing I've noticed is that the prior version allowed me to create recipes that were 15 characters in length, but this version appears to have a minimum of 16. Unfortunately, I saved two recipes with 15 characters, and now I can't recreate them. If I remember correctly, I also noticed that the earlier version would let me manually edit the "internal version of the recipe," e.g. by typing a new value in the "lengthinChars" parameter, but this version seems to no longer allow that.

### PR DESCRIPTION
**Pull request by angelix**
Fix #150

### CODE REVIEW

File: api/src/main/java/org/dicekeys/api/DerivationRecipe.kt [hunk 1]
```diff
@@ -1,52 +1,42 @@
...
+        put("lengthInBytes", lengthInBytes)
+    }
 }
 
-private fun augmentRecipeJson(template: DerivationRecipe, sequenceNumber: Int, lengthInChars: Int, lengthInBytes: Int): String {
-    var recipeJson = template.recipeJson
-    if (template.type == DerivationOptions.Type.Password && lengthInChars > 0) {
-        recipeJson = addLengthInCharsToDerivationOptionsJson(recipeJson, lengthInChars)
+fun JsonObjectBuilder.addSequenceNumberToDerivationOptionsJson(sequenceNumber: Int){
+    if(sequenceNumber > 1){
```
Comment (UppaJung) [inline api/src/main/java/org/dicekeys/api/DerivationRecipe.kt:37]: If the object has a sequenceNumber of 2, and this is called with sequenceNumber of 1, will this remove the old sequenceNumber as it should?

File: api/src/main/java/org/dicekeys/api/DerivationRecipe.kt [hunk 2]
```diff
@@ -1,52 +1,42 @@
...
+    if(lengthInChars > 0){
+        put("lengthInChars", lengthInChars)
+    }
 }
 
-fun addSequenceNumberToDerivationOptionsJson(derivationOptionsWithoutSequenceNumber: String, sequenceNumber: Int): String {
-    if (sequenceNumber == 1) return derivationOptionsWithoutSequenceNumber
-    return addFieldToEndOfJsonObjectString(derivationOptionsWithoutSequenceNumber, "#", sequenceNumber.toString())
+fun JsonObjectBuilder.addLengthInBytesToDerivationOptionsJson(lengthInBytes: Int){
+    if(lengthInBytes > 0){
```
Comment (UppaJung) [inline api/src/main/java/org/dicekeys/api/DerivationRecipe.kt:31]: If lengthInByes is 32, will this remove the lengthInBytes from the object?

File: api/src/main/java/org/dicekeys/api/DerivationRecipe.kt [hunk 3]
```diff
@@ -85,19 +133,26 @@ data class DerivationRecipe constructor(
...
 
-            val allowDomainList = domains.map { """{"host":"*.$it"}""" }
-            val name = domains.joinToString(", ")
-            var recipeJson = """{"allow":[${allowDomainList.joinToString(",")}]}"""
-
-            if (type == DerivationOptions.Type.Password) {
-                recipeJson = addLengthInCharsToDerivationOptionsJson(recipeJson, lengthInChars)
-            }else if (type == DerivationOptions.Type.Secret) {
-                recipeJson = addLengthInBytesToDerivationOptionsJson(recipeJson, lengthInBytes)
+            val jsonObject = buildJsonObject {
```
Comment (UppaJung) [inline api/src/main/java/org/dicekeys/api/DerivationRecipe.kt:136]: What's going on here?  I believe the only time we should be adding a *. to a domain is when the user pastes a URL into the recipe builder.

File: app/src/main/java/org/dicekeys/app/RecipeBuilder.kt [hunk 4]
```diff
@@ -13,9 +14,21 @@ class RecipeBuilder constructor(val type: DerivationOptions.Type, val template:
...
-    private var lengthInChars = 0
-    private var lengthInBytes = 0
+    var rawJson: String? = null
+        private set
+
+    var sequence = 1
+        private set
+    var lengthInChars = template?.lengthInChars ?: 0
+        private set
+    var lengthInBytes = template?.lengthInBytes ?: 0
```
Comment (UppaJung) [inline app/src/main/java/org/dicekeys/app/RecipeBuilder.kt:24]: Why is default 0?

File: app/src/main/java/org/dicekeys/app/RecipeBuilder.kt [hunk 5]
```diff
@@ -64,33 +86,53 @@ class RecipeBuilder constructor(val type: DerivationOptions.Type, val template:
...
+                    DerivationRecipe.createCustomPurposeRecipe(
+                        type = type,
+                        purpose = purpose ?: "",
+                        sequenceNumber = sequence,
+                        lengthInChars = lengthInChars,
+                        lengthInBytes = lengthInBytes
+                    )
+                }
+                else -> {
+                    DerivationRecipe.createCustomOnlineRecipe(
```
Comment (UppaJung) [inline app/src/main/java/org/dicekeys/app/RecipeBuilder.kt:119]: Not sure what this does.  Comments at each step of when would be useful!

File: app/src/main/java/org/dicekeys/app/fragments/PreferencesFragment.kt [hunk 6]
```diff
@@ -1,19 +1,102 @@
...
+import javax.inject.Inject
+
 
 @AndroidEntryPoint
 class PreferencesFragment : PreferenceFragmentCompat() {
+
+    @Inject
+    lateinit var recipeRepository: RecipeRepository
+
+    private val createDocument =
```
Comment (UppaJung) [inline app/src/main/java/org/dicekeys/app/fragments/PreferencesFragment.kt:33]: Please document function

File: app/src/main/java/org/dicekeys/app/fragments/PreferencesFragment.kt [hunk 7]
```diff
@@ -1,19 +1,102 @@
...
+                            }
+                        }
+                    } catch (e: Exception) {
+                        e.printStackTrace()
+                    }
+                }
+            }
+        }
+
+    private val openDocument =
```
Comment (UppaJung) [inline app/src/main/java/org/dicekeys/app/fragments/PreferencesFragment.kt:56]: Please document function

File: api/src/main/java/org/dicekeys/api/DerivationRecipe.kt (hunk 1 above)
Comment (angelix) [reply api/src/main/java/org/dicekeys/api/DerivationRecipe.kt:42]: Good catch. Changed the implementation and added unit tests.

File: api/src/main/java/org/dicekeys/api/DerivationRecipe.kt (hunk 2 above)
Comment (angelix) [reply api/src/main/java/org/dicekeys/api/DerivationRecipe.kt:36]: Same case as before, added unit tests and fixed

File: app/src/main/java/org/dicekeys/app/RecipeBuilder.kt (hunk 4 above)
Comment (angelix) [reply app/src/main/java/org/dicekeys/app/RecipeBuilder.kt:24]: It's not an optional variable and needs initialization, default value 0 means unused in DerivationRecipe constructor.

File: app/src/main/java/org/dicekeys/app/fragments/PreferencesFragment.kt (hunk 7 above)
Comment (angelix) [reply app/src/main/java/org/dicekeys/app/fragments/PreferencesFragment.kt:66]: Done

File: app/src/main/java/org/dicekeys/app/fragments/PreferencesFragment.kt (hunk 6 above)
Comment (angelix) [reply app/src/main/java/org/dicekeys/app/fragments/PreferencesFragment.kt:37]: Done

File: app/src/main/java/org/dicekeys/app/fragments/PreferencesFragment.kt (hunk 7 above)
Comment (UppaJung) [reply app/src/main/java/org/dicekeys/app/fragments/PreferencesFragment.kt:66]: Not seeing any documentation

**Comment by UppaJung** [comment]
If I understand the current code correctly, it supports a domains field which can either have domain names or URLs in it.  The web app supports only domains, and converts URLs to domains when they are pasted, so that they never appear as URLs in the field.  As such, any domains in the field, whether preceded by a wildcard (*.) or not, should be untouched.  *.example.com should keep with wild card and example.com should stick without one.

If I understand this code correctly, it will strip all wildcards.

**Comment by angelix** [comment]
Most of the changes happened in RecipeBuilder in the following block.

```
    private fun getDomainList(): List<String> {
        return domains
            ?.split(",")
            ?.map { it -> it.trim() } // trim whitespace
            ?.map { it -> it.trim { it == '.' || it == '/' } } // remove leading and trailing chars
            ?.distinct() // prevent duplicate unique values
            ?.filter { it.isNotBlank() }
            ?.map { urlOrDomain ->
                try {
                    getWildcardOfRegisteredDomainFromCandidateWebUrl(urlOrDomain) ?: urlOrDomain
                } catch (e: Exception) {
                    urlOrDomain
                }
            }
            ?.sorted() ?: listOf()
    }
```

So even if we miss a paste event, or a user types a url we are certain that we'll make a conversion from url to wildcard domain.
If a user enters a `*.example.com` or `example.com` it will remain unchanged as `getWildcardOfRegisteredDomainFromCandidateWebUrl` won't be able to parse it as a URL and will return null (the try/catch is not needed), in that case we use the `urlOrDomain` provided by the user.

**Review by UppaJung — APPROVED** [review]

**Pull request merged**